yktrace/src/hwt/mapper.rs

```diff
@@ -271,4 +126,60 @@ impl HWTMapper {
         }
         ret
     }
+
+    /// Obtain the next mapped block.
+    ///
+    /// `last_yielded` must be a reference to the last successfully mapped block returned by the
+    /// iterator (or `None` if this is the first time you are calling `next`).
```

Comment:
Why don't we cache a reference to `last_yielded` rather than forcing it on the user?

Reply:
I tried, but failed to get it through the borrow checker. I can try again if you like.

An earlier version used a cloned block, which is awful since it contains a string.

Reply:
Since we're handing out references, I can't see an obvious lifetime issue with this, but I might be wrong. It does feel like it's worth quickly (re)trying.

Reply:
The problem is, you end up trying to store a reference to something that is immediately about to move.

In the `next()` of the mapper:
```
          ...
          let ret = self.buffered_irblocks.pop().unwrap();                                                                                                                                         
          self.last_yielded = Some(&ret);                                                                                                                                                          
          Some(Ok(ret))
```

This obviously doesn't float with the borrow checker, but did you have a different idiom in mind?

Reply:
Ah, I see that this iterator consumes itself (so to speak). In which case, yes, there's not much we can do here (unless we return something like `Result<&IRBlock, ...>` but that might have other consequences).

Reply:
Indeed. The question then becomes: where do the returned blocks then live? The whole purpose of this change is to make the iterator lazy, but to return a reference, we'd be forcing them to be buffered up and collected somewhere anyway...

We may have to suck it up :(

Reply:
Fair enough.